In [25]:
import numpy as np
import pandas as pd
import os

In [26]:
os.getcwd()

'C:\\Users\\Harusha B\\Documents\\Application of software'

In [27]:
os.chdir("C:\\Users\\Harusha B\\Documents\\Application of software")

In [28]:
os.getcwd()

'C:\\Users\\Harusha B\\Documents\\Application of software'

In [29]:
rocinate = pd.read_excel("Exhibit+2.xlsx")
rocinate

Cars  Sales \n(in 1,000 units)  Price\n(in lakh rupees)  \
0    Rocinante 1                   171.877                      6.1   
1    Rocinante 2                   139.796                      6.1   
2    Rocinante 3                   178.947                      9.9   
3    Rocinante 4                   140.022                      5.8   
4    Rocinante 5                   186.476                     10.0   
5    Rocinante 6                   192.123                      6.5   
6    Rocinante 7                   175.085                      5.5   
7    Rocinante 8                   146.882                      8.4   
8    Rocinante 9                   202.847                      6.6   
9   Rocinante 10                   149.933                      8.8   
10  Rocinante 11                   171.579                      9.5   
11  Rocinante 12                   201.512                      9.2   
12  Rocinante 13                   142.604                      8.1   
13  Rocinante 14                   135.158                      5.4   
14  Rocinante 15                   107.322                      5.6   
15  Rocinante 16                   199.699                      6.3   
16  Rocinante 17                   202.875                      5.8   
17  Rocinante 18                   204.406                      7.7   
18  Rocinante 19                   107.331                      7.2   
19  Rocinante 20                   190.345                      8.4   
20  Rocinante 21                   108.419                      8.4   
21  Rocinante 22                   150.249                      5.7   
22  Rocinante 23                   155.882                      5.5   
23  Rocinante 24                   100.980                      8.4   
24  Rocinante 25                   182.679                      5.6   
25  Rocinante 26                   166.752                      9.9   
26  Rocinante 27                   121.561                      6.2   
27  Rocinante 28                   174.256                      9.9   
28  Rocinante 29                   119.018                      6.5   
29  Rocinante 30                   169.842                      7.1   
30  Rocinante 31                   198.311                      9.6   
31  Rocinante 32                   204.875                      5.8   
32  Rocinante 33                   119.561                      6.2   
33  Rocinante 34                   203.875                      5.9   
34  Rocinante 35                   118.561                      6.1   

    Mileage\n(Km/ltr)  Top speed (Km/hr)  
0                15.8              168.2  
1                12.1              149.6  
2                17.0              173.4  
3                11.6              170.6  
4                17.2              175.0  
5                17.6              173.1  
6                16.0              184.6  
7                13.0              175.7  
8                19.3              166.7  
9                13.3              175.4  
10               15.8              150.9  
11               19.2              140.5  
12               11.7              150.2  
13               11.2              146.2  
14                7.8              151.2  
15               19.0              143.2  
16               19.7              178.9  
17               19.5              168.2  
18                7.5              164.7  
19               18.1              182.7  
20                7.9              165.7  
21               12.6              137.0  
22               12.9              173.9  
23                7.6              165.1  
24               16.7              160.1  
25               15.5              155.6  
26                9.5              155.4  
27               16.3              161.9  
28                9.1              146.8  
29               15.3              149.5  
30               19.1              146.5  
31               19.4              178.9  
32                9.4              1

In [30]:
list(rocinate)

['Cars',
 'Sales \n(in 1,000 units)',
 'Price\n(in lakh rupees)',
 'Mileage\n(Km/ltr)',
 'Top speed (Km/hr)']

In [31]:
from sklearn.model_selection import train_test_split

# Renaming the columns

In [32]:
rocinate.columns = ["Cars", "Sales", "Price", "Mileage", "Top_speed"]

In [33]:
X = rocinate[["Price", "Mileage", "Top_speed"]]
y = rocinate["Sales"]

## Splitting the dataset into 70% training and 30% testing sets

In [34]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

## Initializing and training the Linear Regression model

In [35]:
from sklearn.linear_model import LinearRegression

In [36]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

## Making the predictions on the test set

In [37]:
y_pred = model.predict(X_test)

## Calculating the RMSE

In [38]:
from sklearn.metrics import mean_squared_error

In [39]:
rmse = np.sqrt(mean_squared_error(y_test, y_pred))

## RMSE result

In [40]:
print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")

Root Mean Squared Error (RMSE): 2.28


##  Determining if the model is overfitting or underfitting

### Calculating RMSE on the training set

In [41]:
y_train_pred = model.predict(X_train)
rmse_train = np.sqrt(mean_squared_error(y_train, y_train_pred))

### Calculating RMSE on the test set

In [42]:
y_test_pred = model.predict(X_test)
rmse_test = np.sqrt(mean_squared_error(y_test, y_test_pred))

In [43]:
print(f"Training RMSE: {rmse_train:.2f}")
print(f"Test RMSE: {rmse_test:.2f}")

Training RMSE: 2.23
Test RMSE: 2.28


### The model is neither overfitting nor underfitting; it is well-fitted and generalizes well because the training RMSE (2.23) and test RMSE (2.28) are very close